### PR TITLE
[Client types] make args param optional if the input has no required keys

### PR DESCRIPTION
**Pull request by MonsterDeveloper**
This PR modifies the `ClientRequest` type, so the `args` param is only required if the input has required keys (specified by the validator).

CleanShot 2023-10-24 at 12 20 21@2x

See #1583

### CODE REVIEW

File: deno_dist/client/types.ts [hunk 1]
```diff
@@ -16,9 +16,13 @@ export type ClientRequestOptions<T = unknown> = keyof T extends never
 
 type ClientRequest<S extends Schema> = {
   [M in keyof S]: S[M] extends { input: infer R; output: infer O }
-    ? RemoveBlankRecord<Required<R>> extends never
-      ? (args?: {}, options?: ClientRequestOptions) => Promise<ClientResponse<O>>
-      : (args: R, options?: ClientRequestOptions) => Promise<ClientResponse<O>>
+    ? R extends object
```
Comment (yusukebe) [inline deno_dist/client/types.ts:19]: Can we write it as follows?

```ts
type ClientRequest<S extends Schema> = {
  [M in keyof S]: S[M] extends { input: infer R; output: infer O }
    ? R extends object
      ? HasRequiredKeys<R> extends true
        ? (args: R, options?: ClientRequestOptions) => Promise<ClientResponse<O>>
        : (args?: R, options?: ClientRequestOptions) => Promise<ClientResponse<O>>
      : never
    : never
} & {
  $url: () => URL
}
```

File: src/types.ts [hunk 2]
```diff
@@ -580,6 +580,17 @@ export type ExtractSchema<T> = UnionToIntersection<
   T extends Hono<infer _, infer S, any> ? S : never
 >
 
+export type RequiredKeysOf<BaseType extends object> = Exclude<
```
Comment (yusukebe) [inline src/types.ts:583]: Please place `RequiredKeysOf` and `HasRequiredKeys` in `src/utils/types.ts` as they may be used in other places.

**Review by yusukebe — COMMENTED** [review]
Hi @MonsterDeveloper 

Thanks for the PR. I left some comments.

**Review by yusukebe — APPROVED** [review]
Looks good! I'll merge it. Thanks!

**Comment by yusukebe** [comment]
Oops. @MonsterDeveloper Could you run the command to prettify the code?

```text
npx prettier --write ./src
```

**Comment by MonsterDeveloper** [comment]
@yusukebe done:)

**Comment by yusukebe** [comment]
Oops! Please run `denoify`!

**Comment by MonsterDeveloper** [comment]
Sorry, forgot about `denoifying` each commit:) Should be OK now.

**Comment by yusukebe** [comment]
Perfect! I'll merge it now and release the new version including this later. Thank you very much!!

**Pull request merged**